### PR TITLE
Create Mixin for recipe construction

### PR DESCRIPTION
**Pull request by TheComputerGeek2**
This should help reduce errors found in older servers, such as those found from MassiveCombat.

### CODE REVIEW

File: src/com/massivecraft/massivecore/nms/NmsRecipe12.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.massivecraft.massivecore.nms;
+
+import com.massivecraft.massivecore.MassiveCore;
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.ShapelessRecipe;
+
+import java.util.UUID;
+
+public class NmsRecipe12 extends NmsRecipe
```
Comment (oloflarsson) [inline src/com/massivecraft/massivecore/nms/NmsRecipe12.java]: The `12` at the end does not match any of these:
https://github.com/MassiveCraft/MassiveCore/blob/master/src/com/massivecraft/massivecore/mixin/Mixin.java#L195

Could you figure out when the deprecation was introduced? If there was no hard changes to package names I would propose using `NmsRecipe112R1P` which means `"Minecraft 1.12.0+ [1_12_R1+]"`.

File: src/com/massivecraft/massivecore/nms/NmsRecipe12.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	// -------------------------------------------- //
+	// METHODS
+	// -------------------------------------------- //
+	
+	@Override
+	public ShapelessRecipe createShapeless(ItemStack result)
+	{
+		// Avoid using NamespacedKey.randomKey() as it is marked as depreciated
+		return new ShapelessRecipe(new NamespacedKey(MassiveCore.get(), UUID.randomUUID().toString()), result);
+	}
```
Comment (oloflarsson) [inline src/com/massivecraft/massivecore/nms/NmsRecipe12.java]: Can you provide `public ShapelessRecipe createShapeless(ItemStack result, Plugin plugin, String key)` as well? I think we may want to have non-random keys where possible, yet offer a random solution where maybe recipes are too dynamically generated.

I'm thinking in the future the keys will have meaning, so we may want to allow for proper creation of those.

File: src/com/massivecraft/massivecore/nms/NmsRecipeFallback.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.massivecraft.massivecore.nms;
+
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.ShapelessRecipe;
+
+public class NmsRecipeFallback extends NmsRecipe
```
Comment (oloflarsson) [inline src/com/massivecraft/massivecore/nms/NmsRecipeFallback.java]: I propose that we don't have a fallback. Rather call this one `NmsRecipe17R4P` meaning "Minecraft 1.7.10+ [1_7_R4+]".

File: src/com/massivecraft/massivecore/nms/NmsRecipe12.java (hunk 1 above)
Comment (TheComputerGeek2) [reply src/com/massivecraft/massivecore/nms/NmsRecipe12.java]: https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/5ad2ecf658276b8aae336b0e5bfb894a5151b578#src/main/java/org/bukkit/inventory/ShapelessRecipe.java
It was changed in the 1.12 pre-release 2 commit.

Comment (oloflarsson) [reply src/com/massivecraft/massivecore/nms/NmsRecipe12.java]: Then `NmsRecipe112R1P` is the way to go :)

**Comment by VoidLock92** [comment]
really looking forward to this fix we have been waiting to use this on the server :)

**Comment by oloflarsson** [comment]
Merged!

**Comment by Flavourized** [comment]
When is this fix going to be available??

**Pull request closed without merging**